minio/minio
# Add LDAP IDP Configuration APIs

## Description

Adds support for higher level LDAP configuration commands.

## How to test this PR?

Use https://github.com/minio/minio-iam-testing to setup an LDAP env.

Use `mc` from https://github.com/minio/mc/pull/4296

Some sample commands and outputs:

```
$ ./mc admin idp set myminio ldap           server_addr=localhost:1389                      lookup_bind_dn=cn=admin,dc=min,dc=io           lookup_bind_password=admin           user_dn_search_base_dn=dc=min,dc=io           user_dn_search_filter="(uid=%s)"           group_search_base_dn=ou=swengg,dc=min,dc=io           group_search_filter="(&(objectclass=groupofnames)(member=%d))" 
mc: <ERROR> Unable to set IDP config for 'ldap' to server. Result: LDAP Server Connection Error
Detail: Could not connect to LDAP server: LDAP Result Code 200 "Network Error": EOF
Suggestion: Check:
    (1) server address
    (2) TLS parameters, and
    (3) LDAP server's TLS certificate is trusted by MinIO (when using TLS - highly recommended)
Due to: LDAP Result Code 200 "Network Error": EOF.


$ ./mc admin idp set myminio ldap           server_addr=localhost:1389   server_insecure=on                   lookup_bind_dn=cn=admin,dc=min,dc=io           lookup_bind_password=admin1           user_dn_search_base_dn=dc=min,dc=io           user_dn_search_filter="(uid=%s)"           group_search_base_dn=ou=swengg,dc=min,dc=io           group_search_filter="(&(objectclass=groupofnames)(member=%d))" 
mc: <ERROR> Unable to set IDP config for 'ldap' to server. Result: LDAP Lookup Bind Error
Detail: Error connecting as LDAP Lookup Bind user: LDAP Lookup Bind user invalid credentials error: LDAP Result Code 49 "Invalid Credentials": 
Suggestion: Check LDAP Lookup Bind user credentials and if user is allowed to login
Due to: LDAP Lookup Bind user invalid credentials error: LDAP Result Code 49 "Invalid Credentials":.


$ ./mc admin idp set myminio ldap           server_addr=localhost:1389   server_insecure=on                   lookup_bind_dn=cn=admin,dc=min,dc=io           lookup_bind_password=admin           user_dn_search_base_dn=dc=min,dc=io           user_dn_search_filter="(uid=%s)"           group_search_base_dn=ou=swengg,dc=min,dc=io           group_search_filter="(&(objectclass=groupofnames)(member=%d))" 
Successfully applied new settings.
Please restart your server 'mc admin service restart myminio'.

```

## Types of changes

## Checklist:
